docs/adr/adr-007-monitoring-and-alerting.md

```diff
@@ -0,0 +1,267 @@
+jobs actually run we will configure the [RecoverLostJobsWorker][] to identify
+any of these that need to run and attempt them, this protects from lost work
+problems such as [Sidekiq crashing][sidekiq-reliability].
+
+We expect [SendEmailWorker][] and [DigestEmailGenerationWorker][] jobs to run
+quickly (durations of up to a few seconds) and know that there can be high
+volumes of them to run (both can regularly reach volumes of > 100,000).
+Therefore, we are concerned about a scenario where these consistently fail
+because it could take a long time for sufficient retries to occur
+for an alert to be raised (as each retry is placed at the back of the queue).
+To reflect this concern we will contact engineers if either of these workers
+has not completed one successful run within a time period despite work
+existing (suggested time period is 30 minutes). The [Sidekiq retry
+cadence][sidekiq-retry] for these jobs will need to be updated to ensure they
+always retry within the time period, this is to ensure we don't alert while
+work isn't being attempted.
+
+Similarly, we will contact engineers out-of-hours if we determine that
+[DigestRuns][DigestRun] are not created within a reasonable time frame
+(suggestion is 2 hours). Creating these is a [trivial][digest-run-creation],
+but essential, operation for producing a digest and the lack of these suggests
+the system is broken whether it is busy or not.
+
+We will also monitor that runs of [RecoverLostJobsWorker][] and
+[MetricsCollectionWorker][] are succeeding at least once within a time frame
+(suggestion is 1 hour). These jobs should continue running even when the
+system is busy and the lack of them running indicates that we no longer have
+confidence the system is monitored or able to recover. Thus, an alert will be
+created that contacts engineers out-of-hours should these not succeed in a
+time frame.
+
+For Sidekiq workers that don't affect the user noticeable aspects of Email
+Alert API, such as [EmailDeletionWorker][], we won't raise critical alerts if
+these are failing. Instead we will raise a warning if any Sidekiq workers
+exhaust all their retries and are placed in the [dead set][sidekiq-dead]. This
+approach was chosen as it was a pragmatic technique to monitor multiple workers
+without adding alerts for individual workers. Workers will have their
+[retry cadence][sidekiq-retry] considered so that this will alert in a
+reasonable time frame of the retries being exhausted (the Sidekiq default of 21
+days is very long).
+
+We will modify the [latency checks for email queues][latency-checks]. We
+will be treating [transactional emails][] different to other latency checks
+as these emails have have an [expectation of a prompt
+delivery][transactional-delivery-expectation] and the lack of them blocks a
+user from progressing with their task. These will raise a warning after 5 mins
+and go critical (with engineers contacted out-of-hours) should the latency
+reach 15 mins.
```

Comment:
Are we planning to call up people if this only applies to a single transactional email which is not sending?

Reply:
It's based on the queue latency and doesn't consider the number of items. In theory it could be possible to have one item in the queue and a latency of 15 minutes but in practice that should be virtually impossible as you'd only have queue latency when there is a high number of items in the queue.

Reply:
My thoughts were around as we have the `send_email_immediate_high` queue which has the same priority as as the transactional queue then could it be feasible that a single transactional email gets stuck behind a load of high priority emails?

Reply:
Oh I had thought the transactional email was a higher priority than immediate high, that is a surprise. It probably should be higher as they are more important. @benthorner do you have thoughts on that? It doesn't really have an impact as we don't expect that queue to grow big.

Even with them being the same priority it can't be blocked by send_email_immediate_high. The priority indicates how frequently the queue is checked so each time Sidekiq needs to find a new job it'll check one of it's queues, and the priority indicates that it'll check send_email_transactional 8 times more frequently than the default queue.  In practice this means that means that even with a huge send_email_immediate_high queue to work through the first items in every other queue will still get worked on rapidly.

The situation that could block this is if all the Sidekiq workers were busy with a slow running task. For example if we had 300 content changes to process, which take 30 mins to process, and they exhausted all workers then we could find all email is blocked. If we were worried about that scenario we'd probably want to run a separate Sidekiq process for email sending.

Reply:
Ah, thank you for the detailed explanation and for clearing up my concern! 

I agree it feels like transactional emails should be deemed as a higher priority even if the priority isn't going to make much of a difference in practice, due to as you say, there is only going to be a handful of them in the queue at any one time.

Reply:
Thanks for spotting @1pretz1.

I don't mind much either way. An argument for keeping it as-is would be that it emphasises how the intuitive interpretation of "priority" does not match the way Sidekiq works. As long as they are processed more quickly than most other email, I think it's fine to leave it as-is, but I also don't mind if we increase it symbolically.